File: forum/src/main/java/com/example/forum/dto/UserDto.java
package com.example.forum.dto;

import com.example.forum.domain.UserAccount;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.*;

public class UserDto {
    @Getter
    @NoArgsConstructor(access = AccessLevel.PROTECTED)
    public static class SignUp {
        @NotNull private String userId;
        @NotNull @Setter private String userPassword;
        @Email private String email;
        @Setter private String nickname;
        @Setter private String memo;
        @Setter private String role;

        @Builder
        public SignUp(String userId, String userPassword, String email, String nickname, String memo, String role) {
            this.userId = userId;
            this.userPassword = userPassword;
            this.email = email;
            this.nickname = nickname;
            this.memo = memo;
            this.role = role;
        }

        public UserAccount toEntity() {
            return UserAccount.of(userId, userPassword, email, nickname, memo, role);
        }
    }

    @Getter
    @NoArgsConstructor(access = AccessLevel.PROTECTED)
    public static class UserGetReq {
        @NotEmpty private String userId;

        public UserGetReq(String userId, String role) {
            this.userId = userId;
        }
    }

    @Getter @Setter
    @NoArgsConstructor(access = AccessLevel.PROTECTED)
    public static class UserLoginReq {
        @NotEmpty private String userId;
        @NotEmpty private String password;
    }

    @Getter @Setter
    @NoArgsConstructor(access = AccessLevel.PROTECTED)
    public static class ChangeMemoReq {
        @NotEmpty private String userId;
        @NotEmpty private String memo;
    }

    @Getter
    public static class UserGetRes {
        private final String userId;
        private final String email;
        private final String nickname;
        private final String memo;

        @Builder
        public UserGetRes(String userId, String email, String nickname, String memo) {
            this.userId = userId;
            this.email = email;
            this.nickname = nickname;
            this.memo = memo;
        }

        public static UserGetRes from(UserAccount entity) {
            return new UserGetRes(
                    entity.getUserId(),
                    entity.getEmail(),
                    entity.getNickname(),
                    entity.getMemo()
            );
        }
    }
}

File: migration-to-kotlin/src/main/kotlin/com/example/practice/dto/UserDto.kt
package com.example.practice.dto

import com.example.practice.domain.UserAccount
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotNull

class UserDto {
    data class SignUp(
        @field:NotNull var userId : String,
        @field:NotNull var userPassword : String,
        @field:Email var email : String?,
        var nickname : String?,
        var memo : String?
    ) {
        internal fun toEntity() : UserAccount {
            return UserAccount(
                userId = userId,
                userPassword = userPassword,
                email = email ?: "No Email",
                nickname = nickname ?: "No NickName",
                memo = memo ?: "No Memo"
            )
        }
    }

    data class Res(
        val userId : String,
        val email : String,
        val nickname : String,
        val memo : String
    ) {
        constructor(entity : UserAccount) : this(
            userId = entity.userId,
            email = entity.email,
            nickname = entity.nickname,
            memo = entity.memo
        )
    }
}